Lddmm registration

DevinCrowley opened the pull request:
## Description

<!-- If this is a bug-fix or enhancement, state the issue # it closes -->
<!-- If this is a new feature, reference what paper it implements. -->

Add Large Deformation Diffeomorphic Metric Mapping (LDDMM) to the registration module.

This implements the algorithm in:
Computing Large Deformation Metric Mappings via Geodesic Flows of Diffeomorphisms -
https://doi.org/10.1023/B:VISI.0000043755.93987.aa

Some features which are not enabled by default implement the work in:
3D Mapping of Serial Histology Sections with Anomalies Using a Novel Robust Deformable Registration Algorithm -
https://doi.org/10.1007/978-3-030-33226-6_18


## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [x] Unit tests
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->
- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

DevinCrowley committed:
Add files _lddmm.py and _lddmm_utilities to skimage/registration and corresponding tests in the tests folder, update the __init__.py file

DevinCrowley committed:
Update initial_affine input default value from identity to None

DevinCrowley committed:
Update initial_affine default value to None in the user function lddmm_register as well

DevinCrowley committed:
Update skimage/data/__init__.py to cite the Allen Institute in the allen_mouse_brain_atlas function

DevinCrowley committed:
Remove xyz from function and variable names

DevinCrowley committed:
Add resample function in _lddmm_utilities.py, integrate it into _lddmm.py, write a test for it, and import it in skimage/registration/__init__.py

DevinCrowley committed:
Add plot_lddmm.py file to registration examples

DevinCrowley committed:
Update formatting on all lddmm python files and corresponding tests

DevinCrowley committed:
Move allen_mouse_brain_atlas and cleared_mouse_brain from plot_scientific.py to plot_specific.py in skimage.data

DevinCrowley committed:
Update comment in plot_specific.py

DevinCrowley committed:
Update plot_lddmm.py with downsampling and appropriate registration parameters

DevinCrowley committed:
Adjust plot_lddmm.py

DevinCrowley committed:
Update __init__.py

Add newline at end of file

DevinCrowley committed:
Update _lddmm.py

Remove unused import, skimage.transform.resize

DevinCrowley committed:
Update docstrings in _lddmm_utilities.py

DevinCrowley committed:
Merge branch 'lddmm-registration' of https://github.com/neurodata/scikit-image into lddmm-registration

DevinCrowley committed:
Update _lddmm.py

DevinCrowley committed:
Update _lddmm.py with initial draft of new spatially-varying-contrast-map feature

DevinCrowley committed:
Update plot_lddmm.py example with track_progress flag

DevinCrowley committed:
Update _lddmm.py and _lddmm_utilities.py with functioning new spatially_varying_contrast_map feature and untested n-dimensional behavior

DevinCrowley committed:
Replace track_progress with track_progress_every_n in _lddmm.py

DevinCrowley committed:
Fix error with _Lddmm.low_pass_filter

DevinCrowley committed:
Update _lddmm.py with Gauss Newton in the _compute_affine_inv_gradient method

DevinCrowley committed:
Update comment in tests/test_lddmm_utilities.py

DevinCrowley committed:
Update _lddmm.py and tests/test_lddmm.py for testing convenience, not yet complete

DevinCrowley committed:
Update _lddmm.py and test_lddmm.py, not yet complete

DevinCrowley committed:
Update _lddmm.py and test_lddmm.py, all tests pass

DevinCrowley committed:
Update _lddmm.py and tests/test_lddmm.py with new rigid-affine phase and update the example plot_lddmm.py, code works, tests pass, this is not a drill

DevinCrowley committed:
Merge branch 'master' of https://github.com/scikit-image/scikit-image into lddmm-registration

DevinCrowley committed:
Remove completed TODO comment

DevinCrowley committed:
Update data/_registry.py by adding allen_mouse_brain_atlas and cleared_mouse_brain to the registry dictionary

DevinCrowley committed:
Update allen_mouse_brain_atlas and cleared_mouse_brain in data/__init__.py

DevinCrowley committed:
Revert modification of data/_registry.py in attempt to fix integration issues

DevinCrowley committed:
Remodify data/_registry.py to include allen_mouse_brain_atlas and cleared_mouse_brain in the registry dictionary

DevinCrowley committed:
Add fixed_affine_scale argument in _lddmm.py and adjust comment in _lddmm_utilities.py

DevinCrowley committed:
Update _lddmm.py with new lddmm_transform_points user-function, rename apply_lddmm to lddmm_transform_image, change signature of _lddmm_utilities._validate_resolution

DevinCrowley committed:
Update _transform_points argument ingestion to allow more forms for the points argument

DevinCrowley committed:
Update _lddmm.py

DevinCrowley committed:
Fix error in test_lddmm_utilities.py

DevinCrowley committed:
Pass all tests in tests/test_lddmm.py and tests/test_lddmm_utilities.py

DevinCrowley committed:
Update lddmm docstrings from Google to NumPy format, implement infrastructure for multiscale functionality, not yet complete

DevinCrowley committed:
Update exception handling to use raise from where appropriate

DevinCrowley committed:
Refactor tests into test classes with one class per function tested and one test function per test

DevinCrowley committed:
add _lddmm_utilities sinc_resample draft

DevinCrowley committed:
Update lddmm with fully drafted multiscale functionality and tests

DevinCrowley committed:
Update test_lddmm.py with throttled iterations on certain identity tests for speed

DevinCrowley committed:
Replace interpn with RegularGridInterpolator because there is a dumb expense in interpn before it calls RegularGridInterpolator

DevinCrowley committed:
Import sinc_resample in skimage/registration/__init__.py

DevinCrowley committed:
Revert from using RegularGridInterpolator back to interpn

DevinCrowley committed:
Fix error in _update_affine method in _lddmm.py

DevinCrowley committed:
Update multiscale test

DevinCrowley committed:
Hardcode artifact priors, a hack to be corrected

DevinCrowley committed:
Update multiscale to convert between spatially_varying_contrast_map contrast_coefficients cases

DevinCrowley committed:
Fix contrast_coefficient initialization in multiscale

DevinCrowley committed:
Update _lddmm.py with fourier filter corrections and a new argument, preconditioner_smooth_length

DevinCrowley committed:
Correct fourier filter modifications

DevinCrowley committed:
Correct fourier filter modifications

DevinCrowley committed:
Update fourier filters in _lddmm.py

DevinCrowley committed:
Update _lddmm.py

DevinCrowley committed:
Rewrite spatially varying contrast map computation in _compute_contrast_map, add contrast_smooth_length parameter, and rename smooth_length and preconditioner_smooth_length to velocity_smooth_length and preconditioner_velocity_smooth_length

DevinCrowley committed:
Update spatially_varying_contrast_map code

DevinCrowley committed:
Clean up _compute_weights and add padding to values in the assignment of error_at_t in _compute_velocity_fields_gradient

DevinCrowley committed:
Fix error

DevinCrowley committed:
Fix error

DevinCrowley committed:
Fix error

DevinCrowley committed:
Load allen_mouse_brain_atlas and cleared_mouse_brain in data/__init__.py with np.load rather than _load

DevinCrowley committed:
Load allen_mouse_brain_atlas and cleared_mouse_brain in data/__init__.py with _fetch nested within rather than np.load

DevinCrowley committed:
Prepend self. to preconditioner_velocity_smooth_length in assignment of self.preconditioner_low_pass_filter

DevinCrowley committed:
Update transform_image with new output_shape argument

DevinCrowley committed:
Fix error

DevinCrowley committed:
Fix error

DevinCrowley committed:
Add parameters for sigma_background, artifact_prior, background_prior, and maximum_velocity_fields_update, and rename check_artifacts parameter to classify_and_weight_voxels

DevinCrowley committed:
Fix error

DevinCrowley committed:
Fix error

DevinCrowley committed:
Fix error

DevinCrowley committed:
Update docstring

DevinCrowley committed:
Update docstring

DevinCrowley committed:
Remove deprecated contrast code, remove contrast_tolerance, rename contrast_maxiter to contrast_iterations, and update docstring

DevinCrowley committed:
Update plot_lddmm.py example

DevinCrowley committed:
Change default for sigma_regularization to np.inf

DevinCrowley committed:
Restructure output and add map_coordinates_ify parameter

DevinCrowley committed:
Update test_lddmm.py

DevinCrowley committed:
Remove calibrate parameter

DevinCrowley committed:
Update _lddmm.py

DevinCrowley committed:
Update plot_lddmm.py example to demonstrate new interfact compliant with scipy.ndimage.map_coordinates

DevinCrowley committed:
Update comment in plot_lddmm.py

DevinCrowley committed:
Make new outputs work for arbitrary dimensions

DevinCrowley committed:
Update test_lddmm.py to use new output format and remove tests for deprecated apply functions

DevinCrowley committed:
Move generate_position_field from _lddmm.py to _lddmm_utilities.py, update tests accordingly, remove deprecated apply functions from _lddmm.py, and update lddmm_register docstring example

DevinCrowley committed:
Fix test imports and remove deprecated apply functions from __init__.py

DevinCrowley committed:
Begin style overhaul and remove trailing whitespace

DevinCrowley committed:
Format files with Black, adjust comments to match, remove track_progress_every_n parameter and functionality, noting that some lines cannot be reduced below 80 characters with current variable names

DevinCrowley committed:
Rename template to reference_image and target to moving_image despite that being misleading and inferior

DevinCrowley committed:
Apply style update to example plot_lddmm.py

DevinCrowley committed:
Rename classify_and_weight_voxels to artifact_and_background_classification

DevinCrowley committed:
Remove completed todo comments and rename resolution to spacing

DevinCrowley committed:
Resolve merge conflict in skimage/data/_registry.py

DevinCrowley committed:
Update plot_lddmm.py example with modified parameter names

DevinCrowley committed:
Update import style in plot_lddmm.py

DevinCrowley committed:
Add doc/examples/registration/plot_lddmm.ipynb notebook and corresponding checkpoint for easy viewing of the plot_lddmm.py file with outputs

DevinCrowley committed:
Reduce all lines to strictly 79 characters or less

DevinCrowley committed:
Add docstrings to plot_lddmm.py example

DevinCrowley committed:
Merge upstream

DevinCrowley committed:
Remove deprecated test calls in test files

DevinCrowley committed:
Fix reshape error

DevinCrowley committed:
Update data/tests/test_data.py with test_allen_mouse_brain_atlas and test_cleared_mouse_brain tests

DevinCrowley committed:
Constrain line lengths in plot_specific.py

DevinCrowley committed:
Add reject_nans kwarg to _validate_scalar_to_multi in _lddmm_utilities.py and add corresponding test

DevinCrowley committed:
Heavily adjust the lddmm_register signature and rewrite its docstring

DevinCrowley committed:
Update line lengths

DevinCrowley committed:
Fix error

DevinCrowley committed:
Unpack multiscale-relevant kwargs in lddmm_register

DevinCrowley committed:
Fix error

DevinCrowley committed:
Pull data directly into plot_lddmm.py example

DevinCrowley committed:
Remove images from skimage.data

DevinCrowley committed:
Merge branch 'master' of https://github.com/scikit-image/scikit-image into lddmm-registration

DevinCrowley committed:
Remove deprecated plots from plot_specific.py

DevinCrowley committed:
Add pynrrd to requirements/docs.txt

DevinCrowley committed:
Adjust docstring for lddmm_register

DevinCrowley committed:
Fix example in lddmm_register docstring and correct test_reject_nans in test_lddmm_utilities.py

DevinCrowley committed:
Adjust lddmm_register call style in plot_lddmm.py

DevinCrowley committed:
Test docstring

DevinCrowley committed:
Merge upstream master

DevinCrowley committed:
Test docstring

DevinCrowley committed:
Test docstrings

DevinCrowley committed:
Fix docstrings

DevinCrowley committed:
Fix example in lddmm_register docstring

DevinCrowley committed:
Merge branch 'master' of https://github.com/scikit-image/scikit-image into lddmm-registration

DevinCrowley committed:
Add References section to lddmm_register docstring and rename lddmm_register to diffeomorphic_metric_mapping

DevinCrowley committed:
Merge branch 'master' of https://github.com/scikit-image/scikit-image into lddmm-registration

DevinCrowley committed:
Change ax.axis(False) to ax.axis('off') in plot_lddmm.py because Travis is being a little punk

DevinCrowley committed:
Raise error in diffeomorphic_metric_mapping if contrast_order is provided as 0

DevinCrowley committed:
Merge branch 'master' of https://github.com/scikit-image/scikit-image into lddmm-registration

DevinCrowley committed:
Update skimage/registration/_lddmm.py

Replace try except ingestion of multiscales with list(np.atleast_1d(multiscales))

Co-authored-by: Emmanuelle Gouillart <emma@plot.ly>

DevinCrowley committed:
Unindent example in docstring

DevinCrowley committed:
Merge branch 'lddmm-registration' of https://github.com/neurodata/scikit-image into lddmm-registration

DevinCrowley committed:
Remove print statements from test_lddmm.py

DevinCrowley committed:
Replace use of dict and map with dictionary comprehension

DevinCrowley committed:
Refactor the construction of multiscale_lddmm_kwargs

DevinCrowley committed:
Remove unused resample import in _lddmm.py

DevinCrowley committed:
Nest np.array inside np.atleast_1d in the multiscales ingestion of diffeomorphic_metric_mapping to avoid a pending ragged ndarray deprecation in numpy and because Travis is still a little punk

DevinCrowley committed:
Remove deprecated files allen_mouse_brain_atlas.npy and cleared_mouse_brain.npy from skimage/data

DevinCrowley committed:
Merge branch 'master' of https://github.com/scikit-image/scikit-image into lddmm-registration

DevinCrowley committed:
Update references for the data in the plot_lddmm.py example

DevinCrowley committed:
Resolve conflict in requirements/docs.txt

pep8speaks commented:
Hello @DevinCrowley! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`doc/examples/registration/plot_lddmm.py`](https://github.com/scikit-image/scikit-image/blob/16a8a70a899893005d2b31d93be7d505a51700e6/doc/examples/registration/plot_lddmm.py):

> [Line 45:80](https://github.com/scikit-image/scikit-image/blob/16a8a70a899893005d2b31d93be7d505a51700e6/doc/examples/registration/plot_lddmm.py#L45): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (146 > 79 characters)
> [Line 54:80](https://github.com/scikit-image/scikit-image/blob/16a8a70a899893005d2b31d93be7d505a51700e6/doc/examples/registration/plot_lddmm.py#L54): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (97 > 79 characters)
> [Line 422:80](https://github.com/scikit-image/scikit-image/blob/16a8a70a899893005d2b31d93be7d505a51700e6/doc/examples/registration/plot_lddmm.py#L422): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 423:80](https://github.com/scikit-image/scikit-image/blob/16a8a70a899893005d2b31d93be7d505a51700e6/doc/examples/registration/plot_lddmm.py#L423): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 428:80](https://github.com/scikit-image/scikit-image/blob/16a8a70a899893005d2b31d93be7d505a51700e6/doc/examples/registration/plot_lddmm.py#L428): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)



* In the file [`skimage/registration/_lddmm.py`](https://github.com/scikit-image/scikit-image/blob/16a8a70a899893005d2b31d93be7d505a51700e6/skimage/registration/_lddmm.py):

> [Line 188:80](https://github.com/scikit-image/scikit-image/blob/16a8a70a899893005d2b31d93be7d505a51700e6/skimage/registration/_lddmm.py#L188): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)





* In the file [`skimage/registration/tests/test_lddmm_utilities.py`](https://github.com/scikit-image/scikit-image/blob/16a8a70a899893005d2b31d93be7d505a51700e6/skimage/registration/tests/test_lddmm_utilities.py):

> [Line 494:66](https://github.com/scikit-image/scikit-image/blob/16a8a70a899893005d2b31d93be7d505a51700e6/skimage/registration/tests/test_lddmm_utilities.py#L494): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\['
> [Line 701:1](https://github.com/scikit-image/scikit-image/blob/16a8a70a899893005d2b31d93be7d505a51700e6/skimage/registration/tests/test_lddmm_utilities.py#L701): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 0

DevinCrowley commented:
@grlee77 @jni @emmanuelle 
Per @grlee77's suggestion I've recreated the [Lddmm registration PR #4390](https://github.com/scikit-image/scikit-image/pull/4390) here. This seems to appease the GitHub as I can now allow edits by maintainers.

grlee77 committed:
DOC: fix reference links in LDDMM example

grlee77 committed:
pep8 fixes

grlee77 commented:
Okay, I pushed some minor fixes to verify that it is working now. I will revisit again soon with some other changes I had been unable to push previously.

grlee77 committed:
use internal fft module

This will prioritize the scipy.fft module introduced in SciPy 1.4 when available

grlee77 committed:
preserve precision for float32 inputs

grlee77 committed:
replace scipy.interpolate.interpn with scipy.ndimage.map_coordinates

grlee77 committed:
use normalized root-mean-square error metric in LDDMM tests

grlee77 committed:
add plt.show() to LDDMM example

grlee77 commented on skimage/registration/tests/test_lddmm.py:
```diff
@@ -0,0 +1,377 @@
+            coordinates=lddmm_output.reference_image_to_moving_image_transform,
+        )
+
+        assert np.allclose(
+            deformed_reference_image, moving_image, rtol=rtol, atol=atol
+        )
+        assert np.allclose(
+            deformed_moving_image, reference_image, rtol=rtol, atol=atol
+        )
```
This `np.allclose` method with `atol=1 - 1e-9` seems strange to me. The input images are in the range [0, 1] so the tolerance is nearly the full intensity range?

Two other possibilities for comparison are `skimage.metrics.normalized_mutual_information` or `skimage.metrics.normalized_root_mse`. `normalized_root_mse` seems okay as a comparison for these datasets where the intensities of the moving and reference images match.

grlee77 commented:
In 31d4bfc I changed the implementation to use `scipy.ndimage.map_coordinates` rather than `scipy.interpolation.interpn` (it seems to be a bit faster) . As far as I can tell, the example image registration looks the same before and after the change. One motivation for this is potential implementation on the GPU later (we have `scipy.ndimage.map_coordinates` implemented in CuPy).

grlee77 commented:
My primary concern is in playing with the `plot_lddmm.py` demo, the result can converge on unreasonable results if run to more iterations. I didn't spend much time to try and tune settings, but this is a little concerning.

One thing I noticed visually is that there are some very bright isolated pixels in the moving image. I tried modifying the example to crop outliers using:
```Python


def _clip_outliers(img, n_iqr=1.5, nonnegative=True):
    # clip outliers based on a multiple of the interquartile range
    p25, p75 = np.percentile(img, q=[25, 75])
    iqr = p75 - p25
    imin = p25 - n_iqr * iqr
    if nonnegative:
        imin = max(imin, 0)
    imax = p75 + n_iqr * iqr
    return np.clip(img, imin, imax)

reference_image = _clip_outliers(reference_image)
moving_image = _clip_outliers(moving_image)
```

This has not been committed here, but running the example with and without this clipping gave substantially different results.

anorak94 commented:
Hey i think this is a great initiative. Would it be possible to participate in it in some way. There are very few tested pythonic implementations of LDDMM out there. It would be really cool to add SyN and also a way to do landmarks and intensity based registrations together in LDDMM. I know there are ways to do LDDMM using intensity only and using landmarks only but not something that can do both together. Avants et. al. propose a method in their paper Lagrangian frame diffeomorphic image registration: Morphometric comparison of human and chimpanzee cortex. Or something even simpler like including loss term from both intensity and landmarks in the LDDMM algorithm. Please let me know what you think

grlee77 commented:
> Hey i think this is a great initiative. Would it be possible to participate in it in some way. 

Welcome @anorak94. I think there is general interest in having non-linear warping algorithms in scikit-image, but we are a bit short on maintainers with expertise in this area. If you are able to play around with the demo here and/or help review the code and API here from a user perspective that would be useful. I am not personally an expert with LDDMM so any additional feedback from others with expertise in this area is appreciated. Let me know if you need help with how to build scikit-image from this branch so that you can play with the demo locally.

As you may be aware, [there is a SyN implementation in DIPY](https://dipy.org/documentation/1.4.0./examples_built/syn_registration_3d/#example-syn-registration-3d). While reviewing this PR, I had adapted that example to run using the same data as proposed here. For this example, the overlap looks much closer with SyN than with the LDDMM code in the example, but that might just be that the LDDMM parameters are not tuned as well as they could be? 

For the SyN algorithm, last year [I implemented a CUDA-based version for the GPU](https://github.com/dipy/cudipy) that gives an order of magnitude acceleration over the version in DIPY.

vikramc1 commented:
@grlee77 Thank you for your time in reviewing this PR! I've been working with @DevinCrowley and am happy to provide some additional example code registering some example datasets, if you'd like. Please let me know what else is needed to proceed on this PR. 

Thank you!

grlee77 commented:
Hi @vikramc1, my main concerns are:

- the default settings in the current demo do not produce particularly close overlap (see prior comment: https://github.com/scikit-image/scikit-image/pull/5323#issuecomment-830205089). I am not suggesting that it needs to be as good as the SyN result mentioned there, but would hope it can be improved over where it is currently with some tuning.

- There are a ton of parameters, which is nice for power users, but likely confusing/intimidating to the average user. It would help to have a bit more guidance either in the docstrings, example or some new documentation page on how to go about adapting these in real-world use cases. Specifically, if registration is not working as expected which parameters should I start with when tuning?

DevinCrowley commented:
> Hi @vikramc1, my main concerns are:
> 
> * the default settings in the current demo do not produce particularly close overlap (see prior comment: [Lddmm registration #5323 (comment)](https://github.com/scikit-image/scikit-image/pull/5323#issuecomment-830205089)). I am not suggesting that it needs to be as good as the SyN result mentioned there, but would hope it can be improved over where it is currently with some tuning.
> * There are a ton of parameters, which is nice for power users, but likely confusing/intimidating to the average user. It would help to have a bit more guidance either in the docstrings, example or some new documentation page on how to go about adapting these in real-world use cases. Specifically, if registration is not working as expected which parameters should I start with when tuning?

Hey @grlee77, sorry I've been out of touch with this PR (closed I see) for quite some time. I did want to mention though that I had written up a somewhat more extended demo on using this code from scratch, with the fewest number of parameters that a non-power-user might reasonably expect to touch without any deep knowledge. I don't know if this is of interest anymore but I wanted to at least attach the tutorial I made to this.

If this PR might be revived I'd be glad to hear what that would take. I know there are a few people besides myself with an interest in seeing it completed and who may be able contribute to that end.

DevinCrowley committed:
Add doc/examples/registration/lddmm_tutorial.ipynb

grlee77 commented:
> Hey @grlee77, sorry I've been out of touch with this PR (closed I see) for quite some time.

It is still open! I am happy to continue reviewing if you or your colleagues can resume here

anorak94 commented:
Just a small thing maybe this is already known but the SyN version implemented in ANTs is not the time varying velocity field parameterized version. They optimize for the displacement field transform. Id be happy to share a notebook with a toy implementation of SyN in python

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
